### PR TITLE
Tracer method to add property to root span on Datadog

### LINKED ISSUE #61: Suporte para adicionar tags no SpanRoot
Atualmente `Tracer.addProperty(key, value)` [Datadog] adiciona tag no span corrente, desta forma não é possível usar a tag adicionada nos filtros e Dashboard.
Precisamos avaliar a possibilidade de ter algo como `Tracer.addRootProperty(key, value)`.

### PR DESCRIPTION
**Pull request by johnguerson**
Tried to solve [this issue](https://github.com/GuiaBolso/events-protocol/issues/61)

Close #61

### CODE REVIEW

**Comment by Cleidiano** [comment]
I think we should be consistent across all supported trace engine, if the feature is not support by all engine it shouldn't be added to Tracer and TracerEngine interface.
We need some check to ensure that the `NewRelic.addProperty` implementation have the same behavior that are intent to insert with `addRootProperty`.

**Comment by johnguerson** [comment]
Can you look again @Cleidiano?

**Comment by josealmada** [comment]
Não tenho uma ideia melhor de como fazer isso de cara... haha

File: impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt [hunk 1]
```diff
@@ -20,14 +21,61 @@ open class DatadogTracer : TracerEngine<SpanBuilder> {
...
+                is MutableSpan -> {
+                    val mutableSpan = (span as MutableSpan)
+                    val rootSpan = mutableSpan.localRootSpan
+                    rootSpan?.run {
+                        this.setTag(key, value) as Any
+                    }
+                }
+                else -> span.setTag(key, value)
+            }
+        }
```
Comment (Cleidiano) [inline impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt:55]:
```suggestion
        tracer.activeSpan()?.let { span ->
            if (span is MutableSpan) span.localRootSpan?.run { this.setTag(key, value) as Any }
            else span.setTag(key, value)
        }
```

File: impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt [hunk 2]
```diff
@@ -20,14 +21,61 @@ open class DatadogTracer : TracerEngine<SpanBuilder> {
...
+                        val mutableSpan = (span as MutableSpan)
+                        val rootSpan = mutableSpan.localRootSpan
+                        rootSpan?.run {
+                            this.setTag(key, value) as Any
+                        }
+                    }
+                    else -> span.setTag(key, value)
+                }
+            }
+        }
```
Comment (Cleidiano) [inline impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt:76]:
```suggestion
        if (value == null) return

        tracer.activeSpan()?.let { span ->
            if (span is MutableSpan) span.localRootSpan?.run { this.setTag(key, value) as Any }
            else span.setTag(key, value)
        }
```

File: impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt [hunk 3]
```diff
@@ -48,13 +96,27 @@ open class DatadogTracer : TracerEngine<SpanBuilder> {
         }
     }
 
+    override fun notifyRootError(exception: Throwable, expected: Boolean) {
+        tracer.activeSpan()?.let { span ->
+            if (span is MutableSpan) (span as MutableSpan).localRootSpan.isError = true
+            DatadogUtils.notifyError(span, exception, expected)
+        }
```
Comment (Cleidiano) [inline impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt:103]:
```suggestion
        tracer.activeSpan()?.let { span ->
            if (span is MutableSpan) span.localRootSpan.isError = true
            DatadogUtils.notifyError(span, exception, expected)
        }
```

File: impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt [hunk 4]
```diff
@@ -48,13 +96,27 @@ open class DatadogTracer : TracerEngine<SpanBuilder> {
...
         if (span != null) {
             DatadogUtils.notifyError(span, message, params, expected)
         }
     }
 
+    override fun notifyRootError(message: String, params: Map<String, String?>, expected: Boolean) {
+        tracer.activeSpan()?.let { span ->
+            if (span is MutableSpan) (span as MutableSpan).localRootSpan.isError = true
+            DatadogUtils.notifyError(span, message, params, expected)
+        }
```
Comment (Cleidiano) [inline impl/java/tracing/src/main/kotlin/br/com/guiabolso/tracing/engine/datadog/DatadogTracer.kt:117]:
```suggestion
        tracer.activeSpan()?.let { span ->
            if (span is MutableSpan) span.localRootSpan.isError = true
            DatadogUtils.notifyError(span, message, params, expected)
        }
```

**Pull request merged**